app/code/Magento/Catalog/Model/Product/Image/Process/Queue.php

```diff
@@ -0,0 +1,323 @@
+                $this->assertAndExecute($productId, $products);
+            }
+            usleep(20000);
+            foreach ($this->inProgress as $productId) {
+                if ($this->isResized($productId)) {
+                    unset($this->inProgress[$productId]);
+                }
+            }
+        }
+
+        $this->awaitForAllProcesses();
+
+        return $returnStatus;
+    }
+
+    /**
+     * Check that we are not over max processes and execute
+     *
+     * @param int $productId
+     * @param array $products
+     * @return void
+     */
+    public function assertAndExecute($productId, array & $products)
+    {
+        if ($this->maxProcesses < 2 || (count($this->inProgress) < $this->maxProcesses)) {
+            unset($products[$productId]);
+            $this->execute($productId);
+        }
+    }
+
+    /**
+     * Wait untill all processes are finished
+     *
+     * @return void
+     */
+    public function awaitForAllProcesses()
+    {
+        while ($this->inProgress) {
+            foreach ($this->inProgress as $productId => $product) {
+                if ($this->isResized($product)) {
+                    unset($this->inProgress[$productId]);
+                }
+            }
+            usleep(100000);
+        }
+        if ($this->isCanBeParalleled()) {
+            // close connections only if ran with forks
+            $this->resourceConnection->closeConnection();
+        }
+    }
+
+    /**
+     * @return bool
+     */
+    public function isCanBeParalleled()
+    {
+        return function_exists('pcntl_fork') && $this->maxProcesses > 1;
+    }
+
+    /**
+     * @param int $productId
+     * @return bool true on success for main process and exit for child process
+     * @SuppressWarnings(PHPMD.ExitExpression)
+     */
+    public function execute(int $productId)
+    {
+        $this->lastJobStarted = time();
+
+        if ($this->isCanBeParalleled()) {
+            $pid = pcntl_fork();
+            if ($pid === -1) {
+                throw new \RuntimeException('Unable to fork a new process');
+            }
+
+            if ($pid) {
+                $this->inProgress[$productId] = $productId;
+                $this->processIds[$productId] = $pid;
+                return true;
+            }
+
+            try {
+                /** @var \Magento\Catalog\Model\Product $product */
+                $product = $this->productRepository->getById($productId);
+                $product->getMediaGalleryImages();
+            } catch (NoSuchEntityException $e) {
+                exit();
+            }
+
+            // process child process
+            $this->inProgress = [];
+            $this->imageCache->generate($product);
+            $this->output->write('.');
+            exit();
+        }
+
+        try {
+            /** @var \Magento\Catalog\Model\Product $product */
+            $product = $this->productRepository->getById($productId);
```

Comment:
The lines above are for forked php processes. If there is no forking, these lines will be executed. Forked processes will not get here because of the `exit();`. But from the duplicate code perspective I will make a change here nonetheless. Although it adds a few more `if`'s.